src/trace_ng/log.zig

```diff
@@ -0,0 +1,1048 @@
+const std = @import("std");
+const Level = @import("level.zig").Level;
+const logfmt = @import("logfmt.zig");
+const sig = @import("../sig.zig");
+// TODO Improve import.
+const Channel = @import("../sync/channel.zig").Channel;
+const testing = std.testing;
+const Allocator = std.mem.Allocator;
+const AtomicBool = std.atomic.Value(bool);
+const RecycleFBA = sig.utils.allocators.RecycleFBA;
+
+pub const Config = struct {
+    max_level: Level = Level.debug,
+    allocator: std.mem.Allocator,
+    /// Maximum memory that logger can use.
+    max_buffer: ?u64 = null,
+    kind: LogKind = LogKind.standard,
+};
+
+const INITIAL_LOG_CHANNEL_SIZE: usize = 1024;
+
+const LogKind = enum {
+    standard,
+    testing,
+    noop,
+};
+
+/// A ScopedLogger could either be:
+/// - A StandardErrLogger
+/// - A TestingLogger
+pub fn ScoppedLogger(comptime scope: ?[]const u8) type {
+    return union(LogKind) {
+        const Self = @This();
+        standard: *StandardErrLogger(scope),
+        testing: *TestingLogger(scope),
+        noop: void,
+        pub fn init(config: Config) !Self {
+            switch (config.kind) {
+                .standard => {
+                    return .{ .standard = try StandardErrLogger(scope).init(.{
+                        .allocator = config.allocator,
+                        .max_level = config.max_level,
+                        .max_buffer = config.max_buffer,
+                    }) };
+                },
+                .testing, .noop => {
+                    return .{ .testing = TestingLogger(scope).init(.{
+                        .allocator = config.allocator,
+                        .max_level = config.max_level,
+                        .max_buffer = config.max_buffer,
+                    }) };
+                },
+            }
+        }
+
+        pub fn deinit(self: *const Self) void {
+            switch (self.*) {
+                .standard => |*logger| {
+                    logger.*.deinit();
+                },
+                .testing => |*logger| {
+                    logger.*.deinit();
+                },
+                .noop => {},
+            }
+        }
+
```

Comment:
Do we need these functions? The caller could instead initialize whatever logger implementation they want, and convert it to the interface.

```zig
const std_logger = StdErrLogger.init(...);
const logger = std_logger.logger();
```

This approach is more flexible because it doesn't couple all the logger `init` functions together to have a unified set of dependencies.

Reply:
The idea behind the current approach is to have the caller call one init function (specifying the kind of logger they want) and they good to go ie:

```
const std_logger =  Logger.init(.{.. .kind = LogKind.standard,})
```

Your approach would be two calls, but it might indeed be more flexible since it removes the init from the interface. I'll play around with it and see.